### PR TITLE
Validate constrained tasks

### PR DESCRIPTION
**Pull request by crusaderky**

### CODE REVIEW

**Comment by hendrikmakait** [comment]
https://github.com/dask/distributed/runs/7253057570?check_suite_focus=true#step:11:1931

**Comment by crusaderky** [comment]
It feels like the same issue as #6676. I'm blocking this PR until #6676 is in.

File: distributed/worker_state_machine.py [hunk 1]
```diff
@@ -3037,9 +3036,15 @@ def _validate_task_executing(self, ts: TaskState) -> None:
             assert dep.key in self.data or dep.key in self.actors
 
     def _validate_task_ready(self, ts: TaskState) -> None:
-        assert ts.key in pluck(1, self.ready)
+        if ts.state == "ready":
+            assert ts.key in pluck(1, self.ready)
```
Comment (hendrikmakait) [inline distributed/worker_state_machine.py:3040]:
```suggestion
            assert ts.key in pluck(1, self.ready)
            assert ts.key not in self.constrained
```

File: distributed/worker_state_machine.py [hunk 2]
```diff
@@ -3037,9 +3036,15 @@ def _validate_task_executing(self, ts: TaskState) -> None:
...
+            assert not ts.resource_restrictions
+        else:
+            assert ts.state == "constrained"
+            assert ts.key in self.constrained
```
Comment (hendrikmakait) [inline distributed/worker_state_machine.py:3044]:
```suggestion
            assert ts.key in self.constrained
            assert ts.key not in pluck(1, self.ready)
```

**Review by hendrikmakait — COMMENTED** [review]
Validation looks good to me. I'd love to have additional assertions that ensure mutual exclusivity between `self.constrained` and `self.ready`. Should we put this into `_validate_task_ready` as suggested or add another paragraph to `_validate_state` that ensures mutual exclusivity between the various collections we have on the `TaskState`?

**Review by hendrikmakait — APPROVED** [review]
LGTM, see previous review for additional validation I'd like to be added tracked by a separate issue (#6708). Please skip suggested changes.

File: distributed/worker_state_machine.py (hunk 1 above)
Comment (hendrikmakait) [reply distributed/worker_state_machine.py:3040]: Superseded by suggestions in https://github.com/dask/distributed/pull/6698#pullrequestreview-1034032402

File: distributed/worker_state_machine.py (hunk 2 above)
Comment (hendrikmakait) [reply distributed/worker_state_machine.py:3044]: Superseded by suggestions in https://github.com/dask/distributed/pull/6698#pullrequestreview-1034032402

**Comment by crusaderky** [comment]
> EDIT: I think I'd prefer checking that a tasks is the collection we want it to be in, and then checking for each collection that it only contains tasks in the desired state.

It's not that simple.
This is doable with `Set[TaskState]` and HeapSet[TaskState]`, but not with collections of keys such as `ready` and `constrained`. These collections were using keys because you can't cheaply remove a task from the middle of a heap / deque; they predate `HeapSet`. They should be reimplemented.

**Comment by crusaderky** [comment]
> I'd love to have additional assertions that ensure mutual exclusivity between self.constrained and self.ready

Done

**Comment by hendrikmakait** [comment]
> > EDIT: I think I'd prefer checking that a tasks is the collection we want it to be in, and then checking for each collection that it only contains tasks in the desired state.
> 
> It's not that simple. This is doable with `Set[TaskState]` and HeapSet[TaskState]`, but not with collections of keys such as `ready`and`constrained`. These collections were using keys because you can't cheaply remove a task from the middle of a heap / deque; so a task may not be in the right state once it reaches the top of the heap, or it may not exist anymore at all. `HeapSet` addresses this problem exactly, but this collections predate it. They should be reimplemented on top of it.

Could you write up a ticket for reimplementing these if you have any special handling in mind? I just realized that we do this check (https://github.com/dask/distributed/blob/main/distributed/worker_state_machine.py#L1853) in a few places which should have no effect at all given that we are looking for a `str` in a `list[tuple[tuple[int, ...], str]]`.

**Comment by hendrikmakait** [comment]
> > > I'd love to have additional assertions that ensure mutual exclusivity between self.constrained and self.ready
> > 
> > 
> > Done
> 
> Actually no. This would be disastrous: #6710
> 
> > I just realized that we assert ts.key not in self.ready in a few places which should have no effect
> 
> I fixed all the assertions and commented them out for the same reason.

Ouch. Thanks for filing an issue and adding the `# FIXME`s.

**Pull request merged**